Source/Core/Common/MemoryUtil.cpp

```diff
@@ -66,22 +77,47 @@ void* AllocateExecutableMemory(size_t size)
 // PrepareInstructionStreamForJIT();
 // JITPageWriteDisableExecuteEnable();
 
-//Allows a thread to write to executable memory, but not execute the data.
-void JITPageWriteEnableExecuteDisable(){
+// These functions can be nested, in which case execution will only be enabled
+// after the call to the JITPageWriteDisableExecuteEnable from the top most
+// nesting level. Example:
+
+// [JIT page is in execute mode for the thread]
+// JITPageWriteEnableExecuteDisable();
+//   [JIT page is in write mode for the thread]
+//   JITPageWriteEnableExecuteDisable();
+//     [JIT page is in write mode for the thread]
+//   JITPageWriteDisableExecuteEnable();
+//   [JIT page is in write mode for the thread]
+// JITPageWriteDisableExecuteEnable();
+// [JIT page is in execute mode for the thread]
+
+// Allows a thread to write to executable memory, but not execute the data.
+void JITPageWriteEnableExecuteDisable()
+{
+  if (JITPageWriteNestCounter() == 0)
+  {
 #if defined(_M_ARM_64) && defined(__APPLE__)
-  if (__builtin_available(macOS 11.0, *)) {
-    pthread_jit_write_protect_np(0);
-  }
+    if (__builtin_available(macOS 11.0, *))
+    {
+      pthread_jit_write_protect_np(0);
+    }
 #endif
-
+  }
+  JITPageWriteNestCounter()++;
 }
-//Allows a thread to execute memory allocated for execution, but not write to it.
-void JITPageWriteDisableExecuteEnable(){
+// Allows a thread to execute memory allocated for execution, but not write to it.
+void JITPageWriteDisableExecuteEnable()
+{
```

Comment:
`ASSERT(JITPageWriteNestCounter() > 0);`

Reply:
I added a check for the underflow and put the open bracket on a newline. However, I used PanicAlertFmt instead of assert to match the convention used in the rest of the file.